### DISCUSSION TITLE
Grid

### DISCUSSION THREAD

**Discussion by turolopezsanabria** [SUI-Components/sui-components#1676] [Ideas and new features]
# [Layout Grid](https://www.npmjs.com/package/@s-ui/react-layout-grid)

Ready to work on it, [reported as issue here](https://github.com/SUI-Components/sui-components/issues/1709)

Read the discussion below, the winner is `option C`

**Comment by andresin87** [comment]
# Context
The layout grid component adapts to screen size (responsive), ensuring consistency across layouts. The grid creates visual consistency between layouts while allowing flexibility across a wide variety of designs. Based on our design principles, this component is based on a 12 column grid layout.

In the package you can find:
- LayoutGrid
- LayoutGridItem

```js
<LayoutGrid>
  <LayoutGridItem>...</LayoutGridItem>
  <LayoutGridItem>...</LayoutGridItem>
</LayoutGrid>
```


# ^1.1.0
We start developing [@s-ui/react-layout-grid@1](https://www.npmjs.com/package/@s-ui/react-layout-grid/v/1.1.0). This version is mobile-first, which means that the default:
```scss
.sui-LayoutGrid {
  display: flex;
  flex-wrap: wrap;
  list-style: none;
  margin: 0 -8px;
  width: calc(100% + 16px);
}
@media (min-width: 600px) {
  .sui-LayoutGrid {
    flex-wrap: nowrap;
  }
}
```
- Grid Item is 12/12 columns default width (100%).
- 🖥 :  elements do not wrap
- 📱 : elements wrap


### [Codesandbox DEMO v1](https://codesandbox.io/s/s-ui-react-layout-grid-forked-vnes2)

# ^2.0.0
The default behavior makes user be declarative with the desired bahavior.
- Grid Item is 1 columns default width (1/12 => 8.333%).
- 🖥 :  elements wrap
- 📱 : elements wrap
### [Codesandbox DEMO v2](https://codesandbox.io/s/s-ui-react-layout-grid-forked-l702c)



# The problem
The transition between v1 and v2 is a pain because it forces users to be declarative on each item

## Solutions

We appreciate feedback ❤️  so please let us know which solution or solutions you prefer and also which one might not be considered.

You can use reaction icons to each proposed solution for voting:
- 👍 : like.
- 👎 : dislike.
- 😕 : half-hearted.

You can vote as much as you want. Only defined icons will count.

You can find out more info in this [twitch stream](https://www.twitch.tv/videos/1149302614?t=00h05m59s)

**Reply by andresin87** [reply]
## A.- Let it be
<image src="https://user-images.githubusercontent.com/1674036/133569882-9b0cc235-41b4-40a7-85de-04e7d1326121.png" width="500px"/>
I love v2 as is. In my opinion each system might be as mush specific and deterministic as it can be. If we have a 12 column grid system, each item must be 1/12 width by default and I might define the number of columns used for each range until the next bigger defined size. Doesn't matter how much code i need to repeat, i can cache the default behavior and use them in all cases overrwritting the special cases:

```js
const defaultColumns = {
  xxs: 12,
  s: 6,
  l: 3,
  xl: 2,
  xxl: 1
}

...

return (
  <LayoutGrid>
    <LayoutGridItem {...defaultColumns} />
    <LayoutGridItem {...defaultColumns} l={2} />
    <LayoutGridItem {...defaultColumns} l={2} />
    <LayoutGridItem {...defaultColumns} />
  </LayoutGrid>
)
```

**Reply by andresin87** [reply]
## B.- Back to the future
<image src="https://user-images.githubusercontent.com/1674036/133570061-1eb38664-1b2a-474d-91e6-85397a072aec.png" width="500px" />

I'm comfortable with the v1 behavior. Mobile first it the way i think-design-code. Things might be fullscreen unless i decide to do NOT for mobile devices. I want to delegate the default behavior to the css. Please, come back to the v1 taking the advantages of v2 and publish an v3 with it.

```js
return (
  <LayoutGrid>
    <LayoutGridItem />
    <LayoutGridItem />
    <LayoutGridItem />
    <LayoutGridItem />
  </LayoutGrid>
)
```

max-width: 600px:
· –––––––––––– ·
| *************** |
· –––––––––––– ·
| *************** |
· –––––––––––– ·
| *************** |
· –––––––––––– ·
| *************** |
· –––––––––––– ·

higher:
· –––––––––––– · –––––––––––– · –––––––––––– · –––––––––––– ·
| *************** | *************** | *************** | *************** |
· –––––––––––– · –––––––––––– · –––––––––––– · –––––––––––– ·

**Reply by andresin87** [reply]
## C.- Just the two of us

<a href="https://www.youtube.com/embed/v8oqbWrP1QY" target="_blank">
 <img src="https://user-images.githubusercontent.com/1674036/133615874-76fc9242-5917-4b1c-a0ea-b109793135b3.png" alt="Watch the video" width="500" height="300" border="0" />
</a>

Continuous improvement is my second name. I think we can offer the best of both cases giving an hybrid where you can specify on the parent LayoutGrid the default Items behavior and do not repeat on each item everytime. This makes the js deterministic does not beak any v2 rule but column behavior doesn't need to be defined each time.

```js
return (
  <LayoutGrid xxs={12} s={6} l={3} xl={2} xxl={1}>
    <LayoutGridItem />
    <LayoutGridItem l={2} />
    <LayoutGridItem l={2} />
    <LayoutGridItem />
  </LayoutGrid>
)
```

**Reply by adriadoz** [reply]
The C approach is nice too!
